### DISCUSSION TITLE
Speed Issue

### DISCUSSION THREAD

**Discussion by Baron-von-Riedesel** [dosemu2/dosemu2#1728] [General]
Hello,

I've made an experimental piece of software that installs a "DOS VM" and then runs a flat 32-bit application there.
It's nothing complicated, it's "virtual" because DOS memory is mapped in ( using DPMI v1.0 function 0x509 )
and the VGA screen buffer at offset 0xB8000 is "emulated", that is, it is ordinal memory.

The problem is that it's running pretty slow in DosEmu.

The results of a small test app that does nothing but displaying a "hello" string via DOS:

```
                               strings within 2 secs
------------------------------------------------------------ 
bare DOS                         50,000
Qemu+KVM                         20,000
Qemu                                900
DosEmu+KVM                           50
DosEmu+Emulated                      50
```
I'd like to know why DosEmu is so slow and why KVM does not improve things, compared to Qemu

Btw, I also tried CWSDPMI and DPMIONE, but both are unable to run this little gem ( CWSDPMI claims to be "out of memory", DPMIONE just freezes the machine ).

[hellope.exe.zip](https://github.com/dosemu2/dosemu2/files/9085785/hellope.exe.zip)

**Comment by stsp** [comment]
Well it prints `Hello world` and
exits. No any speed measures.

**Comment by Baron-von-Riedesel** [comment]
> No any speed measures.

Yes, I spared me that. After ESC has been pressed, it displays the number of strings displayed. Since the speed differences are in the range of factor 10-1000, rough estimates are quite sufficient.

It's amazing. The experiment's goal was to run multiple PE binaries simultaneously in DosEmu, all of them "thinking" that they run in a flat, zero-based address space.

**Reply by stsp** [reply]
What kind of Esc if it prints
`Hello world` and exits
immediately?

**Reply by andrewbird** [reply]
@stsp it does work for me(tried -t, -X and -S), though guestimating the timing is rough. I achieved around 80 iterations in 2 secs on my ancient
```
Intel(R) Core(TM)2 Quad CPU    Q6600  @ 2.40GHz
```
with kvm and also without.

**Reply by andrewbird** [reply]
Actually I've now done a proper timing and got 291/20 secs.

**Reply by stsp** [reply]
Sorry, I indeed run the
prev version of that prog,
that was posted to some
other thread.

**Reply by stsp** [reply]
`speed 0` and its blazingly fast.

**Comment by stsp** [comment]
So the problem is simple:
this prog polls kbd via int16
too frequently, and that
throttles. The anti-throttling
logic expects some output,
eg via int10, but this prog
seems to be doing output
to video mem directly, so
no anti-throttling happens.

`speed 0` is the only way
to speed it up, for now.

**Comment by Baron-von-Riedesel** [comment]
Ok, now, with speed 0, it runs - surely not "blazingly" but - sufficiently fast. I wasn't aware of that "kbd polling trap" ( although it's also implemented in WinXP VDM ).

**Reply by stsp** [reply]
You can assume that video
memory is "slow" and use
int10 instead. Then dosemu
will see that the program is
doing something besides
polling kbd, and won't throttle
that aggressively.

**Comment by Baron-von-Riedesel** [comment]
> and use int10 instead.

I tried int 0x10, ah=0F and ah=03. Both calls don't speed-up things. And any call that modifies the VGA video buffer is caught, since I want the VGA device to be "virtual".

**Reply by stsp** [reply]
> I tried int 0x10, ah=0F and ah=03.

These are "get cursor position" and
"get video mode". Highly unlikely that
you can use these for text output.

**Comment by Baron-von-Riedesel** [comment]
> Highly unlikely that you can use these for text output.

I don't want to do text output, I want to escape the "kbd polling trap".

**Reply by stsp** [reply]
> I don't want to do text output,

But as I said, video memory is slow. :)
int10 access is much faster here.

> I want to escape the "kbd polling trap".

Either use int10 for output, or this:
```
mov ax,12h
mov bx,0
int 0e6h
```

**Comment by jschwartzenberg** [comment]
Is this why `dir` in dosemu is slower than in dosbox, especially with many entries?

**Reply by stsp** [reply]
> Is this why dir in dosemu is slower than in dosbox, especially with many entries?

Use `speed 0` to check that out.

**Reply by jschwartzenberg** [reply]
I just checked, doesn't really help. In dosbox-x it's instant, but in dosemu it takes a few seconds with a large directory.

**Reply by stsp** [reply]
Try freecom.

**Reply by jschwartzenberg** [reply]
Indeed freecom is quite a bit faster.

**Comment by Baron-von-Riedesel** [comment]
> But as I said, video memory is slow. :) int10 access is much faster here.

it's the main idea behind this little project NOT to allow BIOS output ( not even in real-mode ) and that's why int 10h is hooked in real-mode.

This slowdown of DosEmu is too aggressive. My advice:
- start with the slowdown but after 2 or 3 seconds.
- any video BIOS call or any write access to VGA memory should end the slowdown.

**Comment by stsp** [comment]
Its too aggressive by default.
You can make it less aggressive with
`speed 2` or 10 - the higher value is,
the less aggressive it is.

How do you propose trap accesses -
write protect video mem just for that?

**Comment by Baron-von-Riedesel** [comment]
> How do you propose trap accesses - write protect video mem just for that?

Perhaps just checking the "Dirty" bits of the PTEs? If you're at it,  why not implementing support for such info in int 31h, ax=506h :))

**Comment by stsp** [comment]
Its not easy to access that bit under
linux. And I fail to see how would you
use it to prevent throttling: collect these
bits from all video pages upon every
int16 call?
Just put `speed 2` to your userhook.bat
or set `$_hogthreshold=(2)`.

**Comment by Baron-von-Riedesel** [comment]
OK, I agree that comparatively complicated "idle detection" routines are the wrong way.

But my first approach, catching calls of the BIOS getcursor/setcursor function ( or even trap the corresponding CRT register access if that's not too hard to do ) should tell that the machine is busy, not idle.

**Comment by stsp** [comment]
Yes, putting reset_idle() calls to crt reg
writes may be acceptable.
Please do that and submit a patch if it
actually helps.

**Comment by Baron-von-Riedesel** [comment]
Ok, I added a reset_idle( 1 ) call to crtemu.c and it makes things a lot better ( factor 20 ).
However, I added the call when index register 0x0f ( cursor pos ) is set, so only text modes might benefit.

```
    case 0x0f:		/* Cursor Location  Low */
      vga.crtc.cursor_location = (u + (vga.crtc.data[0x0e] << 8)) << vga.crtc.addr_mode;
      reset_idle( 1 );
      crtc_deb("CRTC_write_value: Cursor Location = 0x%04x\n", vga.crtc.cursor_location);
      break;
```

Also, I have no idea what the argument for reset_idle() is needed for, testing 0, -1 and 1, finally choosing 1 because it gave the best results...

**Reply by stsp** [reply]
All these args are empirical.
So please add it into all relevant
places (cursor location High etc)
and create a PR.
Could you please do that?
